### DISCUSSION TITLE
ExifThumb

### DISCUSSION THREAD

**Discussion by telenkor** [jim-easterbrook/python-exiv2#16] [Q&A] [answered]
I tried to extract the preview based on the code [from here](https://github.com/jim-easterbrook/python-exiv2/blob/main/examples/thumbnail.py).

```python
file = "/Volumes/0001.NEF"
image = exiv2.ImageFactory.open(file)
image.readMetadata()
exifData = image.exifData()

thumb = exiv2.ExifThumb(exifData)
print('Thumbnail type:', thumb.mimeType())
data = thumb.copy()
thumb_im = PIL.Image.open(io.BytesIO(data.data()))
```
It doesn’t get to the opening in PIL because `thumb.mimeType()` is "", data — 0 bytes.

**Comment by jim-easterbrook** [comment, marked as answer]
Your NEF file probably doesn't have a thumbnail. Most raw formats use "preview" images (at various sizes) instead. The libexiv2 `PreviewManager` class handles those. Here's how I do it in one of my applications: https://github.com/jim-easterbrook/Photini/blob/main/src/photini/exiv2.py#L482

**Reply by telenkor** [reply]
No, the file is fine. It's just that the procedure for extracting previews turned out to be a little more complicated. Thanks to researching Photini's code, I can not only extract the previews, but also control which one to choose.

**Reply by jim-easterbrook** [reply]
Yes, the file is fine. It just doesn't have an Exif thumbnail. There's more info in the exiv2 docs:
https://exiv2.org/doc/classExiv2_1_1ExifThumb.html#details

**Reply by telenkor** [reply]
I have CR3 files and I can’t get previews larger than 1620x1080 pixels from them. Reading [discussions](https://github.com/Exiv2/exiv2/issues/1893) about EXIV2 bugs, I realized that the BMFF bug and CR3 preview were fixed in version 0.27.5 You have version 0.27.7 in exiv2-0.15.0, correct? It turns out that this problem has not been fixed after all?

**Reply by jim-easterbrook** [reply]
You can check what version of libexiv2 you are using with `python3 -m exiv2`.

**Reply by jim-easterbrook** [reply]
Sorry about that terse reply. Yes, the "binary wheels" on PyPI currently include libexiv2 v0.27.7. However, if for some reason you install the source from PyPI then it gets built against your system libexiv2 (assuming you have the necessary build tools and package header files installed). `python3 -m exiv2` should tell you exactly which libexiv2 version python-exiv2 is using.

**Reply by telenkor** [reply]
Don't worry, it's okay! To extract the JPEG miniatures I simply returned to **_rawpy_**. This is due to the fact that, as I understand it, EXIV2 even the latest version 0.28.1 (command line version), does not see the full-size preview in CR3. So my attempt to outsource all the work with EXIV2 metadata was unsuccessful. But **_exiv2_** performs its task of working with text metadata to the extent necessary for me as it should.

**Reply by jim-easterbrook** [reply]
Just for interest - I may have fixed my problem with Canon maker notes being truncated in exiv2 v0.28 (https://github.com/Exiv2/exiv2/pull/2860) so the next release of python-exiv2 might be using v0.28.2.

**Reply by telenkor** [reply]
Cool news. I'll keep an eye on the updates. As soon as it comes out, I will be glad to experiment with this version in my project.

**Reply by jim-easterbrook** [reply]
It depends when exiv2 do their next release. Python-exiv2 v0.16.0 is almost ready for release and I don't want to delay it too long. (You can see what's changed by looking at the v0.16.0 branch on GitHub.)

**Reply by jim-easterbrook** [reply]
I've just released python-exiv2 v0.16.0 (with libexiv2 0.28.1) and python-exiv2 v0.16.1 (with libexiv2 0.27.7). This means that a simple "install latest" will still be on the older libexiv2, but you can try the newer libexiv2 with an explicit `pip install --user exiv2==0.16.0`.

**Reply by telenkor** [reply]
Great news for the New Year! I tried the version with the 0.28.x library. Everything is fine. My patch that corrected the [header of Webp images](https://github.com/jim-easterbrook/python-exiv2/discussions/14#discussioncomment-7668794) is no longer needed.
For users of the virtual environment (PyCharm), install version 16.0 with the following command: `pip install exiv2==0.16.0`
